packages/core/src/external/fhir/bundle/qa.ts

```diff
@@ -45,24 +47,46 @@ export function checkBundleForPatient(
     log(
       `Found ${mismatchingPatientsIds.length} mismatching patients in bundle: ${mismatchingPatientsIds}`
     );
-    throw new MetriportError("Unexpected patient IDs in bundle", undefined, {
+    throw new MetriportError("Unexpected patient ID in bundle", undefined, {
       cxId,
       patientId,
-      mismatchingPatientsIds: mismatchingPatientsIds.join(", "),
+      unexpectedPatientsIds: mismatchingPatientsIds.join(", "),
     });
   }
 
   return true;
 }
 
 export function getPatientIdsFromBundle(bundle: Bundle<Resource>): string[] {
-  const contents = JSON.stringify(bundle);
-  const matches = contents.match(/"Patient\/(.+?)"/g);
-  const uniquePatientIds = uniq(matches).map(getPatientIdFromRef).flatMap(filterTruthy);
+  const resources = (bundle.entry ?? []).flatMap(entry => entry.resource ?? []);
+  return [
+    ...getPatientIdsFromPatient(resources),
+    ...getPatientIdsFromContained(resources),
+    ...getPatientIdsFromReferences(resources),
+  ];
+}
+
+export function getPatientIdsFromPatient(resources: Resource[]): string[] {
+  const patientResources = resources.filter(isPatient);
+  const patientIdsFromPatients = patientResources.map(p => p.id).flatMap(filterTruthy);
+  const uniquePatientIds = uniq(patientIdsFromPatients);
   return uniquePatientIds;
 }
 
-export function getPatientIdFromRef(ref: string): string | undefined {
-  const id = ref.split("/")[1];
-  return id ? id.replace(/"/g, "") : undefined;
+export function getPatientIdsFromContained(resources: Resource[]): string[] {
+  const containedResources = resources.flatMap(r => ("contained" in r ? r.contained ?? [] : []));
+  const patientIdsFromContained = containedResources
+    .filter(isPatient)
+    .map(p => p.id)
+    .flatMap(filterTruthy);
+  const uniquePatientIds = uniq(patientIdsFromContained);
+  return uniquePatientIds;
+}
+
+export function getPatientIdsFromReferences(resources: Resource[]): string[] {
+  const references = getPatientReferencesFromResources(resources);
+  const patientReferences = references.filter(isPatientReference);
+  const patientIdsFromRefs = patientReferences.map(getIdFromReference).flatMap(filterTruthy);
```

Comment:
why not just have getPatientReferencesFromResources return the ids? Doesnt seem efficient to iterate through this list again ...

Reply:
Because the function might be used for something else. An additional `O(n)` on the list of patient references is not going to make or break the API.

And the plan is to create a generic `getReferencesFromResources` that will return the references anyways, so I'd leave as is since it's easier to read and reuse.